incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -783,41 +1135,1198 @@ private void onKafkaAbort(
+                h.value(value);
+            });
+        }
+
+        private void addFiltersHeader(
+            String16FW responseTopic)
+        {
+            final DirectBuffer responseBuffer = responseTopic.value();
+            final int capacity = responseBuffer.capacity();
+
+            int offset = 0;
+            int matchAt = 0;
+            while (offset >= 0 && offset < capacity && matchAt != -1)
+            {
+                matchAt = indexOfByte(responseBuffer, offset, capacity, SLASH_BYTE);
+                if (matchAt != -1)
+                {
+                    addHeader(helper.kafkaReplyFilterHeaderName, responseBuffer, offset, matchAt - offset);
+                    offset = matchAt + 1;
+                }
+            }
+            addHeader(helper.kafkaReplyFilterHeaderName, responseBuffer, offset, capacity - offset);
+        }
+
+        private void addHeader(
+            OctetsFW key,
+            MqttPayloadFormatFW format)
+        {
+            String16FW value = format.get() == MqttPayloadFormat.BINARY ? binaryFormat : textFormat;
+            addHeader(key, value);
+        }
+
+        private void addHeader(
+            OctetsFW key,
+            String16FW value)
+        {
+            DirectBuffer buffer = value.value();
+            kafkaHeadersRW.item(h ->
+            {
+                h.nameLen(key.sizeof());
+                h.name(key);
+                h.valueLen(value.length());
+                h.value(buffer, 0, buffer.capacity());
+            });
+        }
+
+        private void addHeader(
+            OctetsFW key,
+            DirectBuffer buffer,
+            int offset,
+            int length)
+        {
+            kafkaHeadersRW.item(h ->
+            {
+                h.nameLen(key.sizeof());
+                h.name(key);
+                h.valueLen(length);
+                h.value(buffer, offset, length);
+            });
+        }
+
+        private void addHeader(String16FW key, String16FW value)
+        {
+            DirectBuffer keyBuffer = key.value();
+            DirectBuffer valueBuffer = value.value();
+            kafkaHeadersRW.item(h ->
+            {
+                h.nameLen(key.length());
+                h.name(keyBuffer, 0, keyBuffer.capacity());
+                h.valueLen(value.length());
+                h.value(valueBuffer, 0, valueBuffer.capacity());
+            });
+        }
+    }
+
+    private static int indexOfByte(
+        DirectBuffer buffer,
+        int offset,
+        int limit,
+        byte value)
+    {
+        int byteAt = -1;
+        for (int index = offset; index < limit; index++)
+        {
+            if (buffer.getByte(index) == value)
+            {
+                byteAt = index;
+                break;
+            }
+        }
+        return byteAt;
+    }
+
+    private static boolean isSetWillFlag(
+        int flags)
+    {
+        return (flags & MqttSessionFlags.WILL.value() << 1) != 0;
```

Comment:
Does this behave the way you want, with bit shift happening before bitwise and?

Reply:
Yes. I did it this way so we don't need to change the flags coming from the network protocol, we can just pass it through the mqtt-kafka binding.
WILL_FLAG_MASK = 0b0000_0100;
CLEAN_START_FLAG_MASK = 0b0000_0010;
So in the mqtt idl: CLEAN_START(1), WILL(2)
Actually maybe it'd be better to provide the actual value, so we don't need shifting? So in this case 2 and 4? What do you think?

Reply:
I think it's already good like this. 👍